contracts/AtomicSwap.sol

```diff
@@ -24,103 +24,97 @@ contract AtomicSwapERC20 {
-        address _erc20ContractAddress, 
-        address _withdrawTrader,
-        bytes32 _secretLock, 
-        uint256 _timelock
+        uint256 _openValue, 
+        address _openContractAddress, 
+        uint256 _closeValue, 
+        address _closeTrader, 
+        address _closeContractAddress
     ) 
         public 
         onlyInvalidSwaps(_swapID) 
-    {
+    {   
         require(swapStates[_swapID] == States.INVALID);
-        // Transfer value from the ERC20 trader to this contract.
-        ERC20 erc20Contract = ERC20(_erc20ContractAddress);
-        require(_erc20Value <= erc20Contract.allowance(msg.sender, address(this)));
-        require(erc20Contract.transferFrom(msg.sender, address(this), _erc20Value));
-
         // Store the details of the swap.
         Swap memory swap = Swap({
-            timelock: _timelock,
-            erc20Value: _erc20Value,
-            erc20Trader: msg.sender,
-            erc20ContractAddress: _erc20ContractAddress,
-            withdrawTrader: _withdrawTrader,
-            secretLock: _secretLock,
-            secretKey: new bytes(0)
+            openValue: _openValue,
+            openTrader: msg.sender,
+            openContractAddress: _openContractAddress,
+            closeValue: _closeValue,
+            closeTrader: _closeTrader,
+            closeContractAddress: _closeContractAddress
         });
         swaps[_swapID] = swap;
         swapStates[_swapID] = States.OPEN;
-        emit Open(_swapID, _withdrawTrader, _secretLock);
+
+        emit Open(_swapID, _closeTrader);
     }
 
     function close(
-        bytes32 _swapID, 
-        bytes memory _secretKey
+        bytes32 _swapID
     ) 
         public 
-        onlyOpenSwaps(_swapID)
-        onlyWithSecretKey(_swapID, _secretKey) 
+        onlyOpenSwaps(_swapID) 
+        onlyCloseTrader(_swapID) 
     {
-        // Close the swap.
         Swap memory swap = swaps[_swapID];
-        swaps[_swapID].secretKey = _secretKey;
-        swapStates[_swapID] = States.CLOSED;
 
-        // Transfer the ERC20 funds from this contract to the withdrawing trader.
-        ERC20 erc20Contract = ERC20(swap.erc20ContractAddress);
-        require(erc20Contract.transfer(swap.withdrawTrader, swap.erc20Value));
+        // both parties have enough tokens
+        ERC20 openERC20Contract = ERC20(swap.openContractAddress);
+        ERC20 closeERC20Contract = ERC20(swap.closeContractAddress);
+        require(swap.openValue <= openERC20Contract.allowance(swap.openTrader, address(this)));
+        require(swap.openValue <= openERC20Contract.balanceOf(swap.openTrader));
+        require(swap.closeValue <= closeERC20Contract.allowance(swap.closeTrader, address(this)));
+        require(swap.closeValue <= closeERC20Contract.balanceOf(swap.closeTrader));
+
+        // Transfer the closing funds from the closing trader to the opening trader.
+        require(closeERC20Contract.transferFrom(swap.closeTrader, swap.openTrader, swap.closeValue));
```

Comment:
This will work most of the time, however it might be worth using OpenZepplin's SafeERC20 lib. There are a number of old ERC20 tokens (OMG being one of them!) that don't strictly follow the latest ERC20 standard, so can cause problems when transferring. The SafeERC20 lib handles all those edge cases for you.

Here's an example from the Plasma contracts
https://github.com/omgnetwork/plasma-contracts/blob/ad89d376532e7e21774655135a74229572f2f647/plasma_framework/contracts/src/vaults/Erc20Vault.sol#L42

Reply:
`import "./ERC20.sol";` was replaced by `import "@openzeppelin/contracts/token/ERC20/SafeERC20.sol";`. `transferFrom` was updated to 
```
// Transfer the closing funds from the closing trader to the opening trader.
require(closeERC20Contract.safeTransferFrom(swap.closeTrader, swap.openTrader, swap.closeValue));

// Transfer the opening funds from opening trader to the closing trader.
require(openERC20Contract.safeTransferFrom(swap.openTrader, swap.closeTrader, swap.openValue));
```